### PR TITLE
Consolidate TTS cache logic

### PR DESCRIPTION
**Pull request by chrisveilleux**
## Description
The impetus for this change is an attempt to fix an issue with pre-loaded cache files on devices that use Mimic2.
- Consolidate all the logic dealing with TTS caching into a single module.  
- Replace logic that copies persistent cache to temporary cache with logic that checks both places

## How to test
Check that persistent cache files are not in `/tmp` and still get used when requested. 

## Contributor license agreement signed?
CLA [Y]

### CODE REVIEW

File: mycroft/tts/cache.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        except Exception:
+            LOG.exception("Failed to write {} to cache".format(self.name))
+
+
+class TextToSpeechCache:
+    """Class for all persistent and temporary caching operations."""
+    def __init__(self, tts_config, tts_name, audio_file_type):
+        self.config = tts_config
+        self.tts_name = tts_name
+        self.persistent_cache_dir = Path(tts_config.get("preloaded_cache"))
```
Comment (forslund) [inline mycroft/tts/cache.py:132]: This config doesn't exist for all tts's and throws an error when missing.

File: mycroft/tts/tts.py [hunk 2]
```diff
@@ -351,22 +355,46 @@ def _execute(self, sentence, ident, listen):
...
             else:
-                wav_file, phonemes = self.get_tts(sentence, wav_file)
+                # TODO: this should be changed return the audio data from
+                #  the API call and then to call the add_to_cache method
+                #  of the TTS cache.  But this requires changing the public
+                #  API of the get_tts method in each engine.
+                audio_file, phoneme_file = self._add_sentence_to_cache(
+                    sentence_hash
+                )
+                _, phonemes = self.get_tts(sentence, str(audio_file.path))
```
Comment (forslund) [inline mycroft/tts/tts.py:376]: Isn't it clearer to rely on the return than use the side-effect?

Is the order ideal here.

First we add something to the cache then we try to get / generate the content of that cache. What would happen if the fetch/generation fails. Would there be a cache entry with no data?

**Review by forslund — COMMENTED** [review]
Looks really good. Great to get rid of the copy from persistent cache to the tmp one.

I noted two things (one that is an iseue, and one that may just be me not quite grasping the idea)

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2853)). 
Mycroft logs are also available: [skills.log](https://reports.mycroft.ai/core/PR-2853/logs/skills.log), [audio.log](https://reports.mycroft.ai/core/PR-2853/logs/audio.log), [voice.log](https://reports.mycroft.ai/core/PR-2853/logs/voice.log), [bus.log](https://reports.mycroft.ai/core/PR-2853/logs/bus.log), [enclosure.log](https://reports.mycroft.ai/core/PR-2853/logs/enclosure.log)

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2853)). 
Mycroft logs are also available: [skills.log](https://reports.mycroft.ai/core/PR-2853/logs/skills.log), [audio.log](https://reports.mycroft.ai/core/PR-2853/logs/audio.log), [voice.log](https://reports.mycroft.ai/core/PR-2853/logs/voice.log), [bus.log](https://reports.mycroft.ai/core/PR-2853/logs/bus.log), [enclosure.log](https://reports.mycroft.ai/core/PR-2853/logs/enclosure.log)

File: mycroft/tts/cache.py (hunk 1 above)
Comment (chrisveilleux) [reply mycroft/tts/cache.py:132]: Good catch.  Will fix.

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2853)). 
Mycroft logs are also available: [skills.log](https://reports.mycroft.ai/core/PR-2853/logs/skills.log), [audio.log](https://reports.mycroft.ai/core/PR-2853/logs/audio.log), [voice.log](https://reports.mycroft.ai/core/PR-2853/logs/voice.log), [bus.log](https://reports.mycroft.ai/core/PR-2853/logs/bus.log), [enclosure.log](https://reports.mycroft.ai/core/PR-2853/logs/enclosure.log)

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2853)). 
Mycroft logs are also available: [skills.log](https://reports.mycroft.ai/core/PR-2853/logs/skills.log), [audio.log](https://reports.mycroft.ai/core/PR-2853/logs/audio.log), [voice.log](https://reports.mycroft.ai/core/PR-2853/logs/voice.log), [bus.log](https://reports.mycroft.ai/core/PR-2853/logs/bus.log), [enclosure.log](https://reports.mycroft.ai/core/PR-2853/logs/enclosure.log)

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2853)). 
Mycroft logs are also available: [skills.log](https://reports.mycroft.ai/core/PR-2853/logs/skills.log), [audio.log](https://reports.mycroft.ai/core/PR-2853/logs/audio.log), [voice.log](https://reports.mycroft.ai/core/PR-2853/logs/voice.log), [bus.log](https://reports.mycroft.ai/core/PR-2853/logs/bus.log), [enclosure.log](https://reports.mycroft.ai/core/PR-2853/logs/enclosure.log)

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2853))

**Review by krisgesling — APPROVED** [review]

**Pull request merged**